txn-file: Panic for nil pointer on calling `locate.(*Region).GetMeta`

Stack:

```
[2024/07/09 17:56:05.169 +00:00] [ERROR] [conn.go:1139] ["connection running loop panic"] [keyspaceName=ks1] [keyspaceID=1] [conn=672859334757056925] [gw_conn=] [lastSQL=commit] [err="runtime error: invalid memory address or nil pointer dereference"] [stack="github.com/pingcap/tidb/server.(*clientConn).Run.func1
	/go/src/github.com/pingcap/tidb/server/conn.go:1142
runtime.gopanic
	/usr/local/go/src/runtime/panic.go:914
runtime.panicmem
	/usr/local/go/src/runtime/panic.go:261
runtime.sigpanic
	/usr/local/go/src/runtime/signal_unix.go:861
github.com/tikv/client-go/v2/internal/locate.(*Region).GetMeta
	/go/cache/github.com/iosmanthus/client-go/v2@v2.0.0-alpha.0.20240625062808-7255d6566924/internal/locate/region_cache.go:2267
github.com/tikv/client-go/v2/txnkv/transaction.(*twoPhaseCommitter).beforeExecuteTxnFile
	/go/cache/github.com/iosmanthus/client-go/v2@v2.0.0-alpha.0.20240625062808-7255d6566924/txnkv/transaction/txn_file.go:894
github.com/tikv/client-go/v2/txnkv/transaction.(*twoPhaseCommitter).executeTxnFile
	/go/cache/github.com/iosmanthus/client-go/v2@v2.0.0-alpha.0.20240625062808-7255d6566924/txnkv/transaction/txn_file.go:551
github.com/tikv/client-go/v2/txnkv/transaction.(*KVTxn).Commit
	/go/cache/github.com/iosmanthus/client-go/v2@v2.0.0-alpha.0.20240625062808-7255d6566924/txnkv/transaction/txn.go:510
github.com/pingcap/tidb/store/driver/txn.(*tikvTxn).Commit
	/go/src/github.com/pingcap/tidb/store/driver/txn/txn_driver.go:99
github.com/pingcap/tidb/session.(*LazyTxn).Commit
	/go/src/github.com/pingcap/tidb/session/txn.go:429
github.com/pingcap/tidb/session.(*session).commitTxnWithTemporaryData
	/go/src/github.com/pingcap/tidb/session/session.go:809
github.com/pingcap/tidb/session.(*session).doCommit
	/go/src/github.com/pingcap/tidb/session/session.go:690
github.com/pingcap/tidb/session.(*session).doCommitWithRetry
	/go/src/github.com/pingcap/tidb/session/session.go:943
github.com/pingcap/tidb/session.(*session).CommitTxn
	/go/src/github.com/pingcap/tidb/session/session.go:1070
github.com/pingcap/tidb/session.autoCommitAfterStmt
	/go/src/github.com/pingcap/tidb/session/tidb.go:297
github.com/pingcap/tidb/session.finishStmt
	/go/src/github.com/pingcap/tidb/session/tidb.go:259
github.com/pingcap/tidb/session.runStmt
	/go/src/github.com/pingcap/tidb/session/session.go:2432
github.com/pingcap/tidb/session.(*session).ExecuteStmt
	/go/src/github.com/pingcap/tidb/session/session.go:2262
github.com/pingcap/tidb/server.(*TiDBContext).ExecuteStmt
	/go/src/github.com/pingcap/tidb/server/driver_tidb.go:315
github.com/pingcap/tidb/server.(*clientConn).executePreparedStmtAndWriteResult
	/go/src/github.com/pingcap/tidb/server/conn_stmt.go:308
github.com/pingcap/tidb/server.(*clientConn).executePlanCacheStmt
	/go/src/github.com/pingcap/tidb/server/conn_stmt.go:236
github.com/pingcap/tidb/server.(*clientConn).handleStmtExecute
	/go/src/github.com/pingcap/tidb/server/conn_stmt.go:227
github.com/pingcap/tidb/server.(*clientConn).dispatch
	/go/src/github.com/pingcap/tidb/server/conn.go:1472
github.com/pingcap/tidb/server.(*clientConn).Run
	/go/src/github.com/pingcap/tidb/server/conn.go:1223
github.com/pingcap/tidb/server.(*Server).onConn
	/go/src/github.com/pingcap/tidb/server/server.go:753"]
```

--- a/txnkv/transaction/txn_file.go
+++ b/txnkv/transaction/txn_file.go
@@ -884,11 +884,27 @@ func (c *twoPhaseCommitter) beforeExecuteTxnFile(
 
 	ctx := bo.GetCtx()
 	regionCache := c.store.GetRegionCache()
-	loc, err := regionCache.LocateKey(bo, c.primary())
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to locate primary")
+
+	var region *locate.Region
+	for {
+		loc, err := regionCache.LocateKey(bo, c.primary())
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to locate primary")
+		}
+
+		region = regionCache.GetCachedRegionWithRLock(loc.Region)
+		if region != nil {
+			break
+		}
+
+		err = bo.Backoff(retry.BoRegionMiss, errors.New("cached region not found"))
+		if err != nil {
+			logutil.Logger(ctx).Error("txn file: cached region not found",
+				zap.String("key", kv.StrKey(c.primary())),
+				zap.Stringer("loc", loc))
+			return nil, errors.WithStack(err)
+		}
 	}
-	region := regionCache.GetCachedRegionWithRLock(loc.Region)
 
 	var replicaNumber int64 = 0
 	for _, peer := range region.GetMeta().GetPeers() {

